### PR TITLE
Lemke refactor

### PR DESCRIPTION
**Pull request by tokheim**
I wanted to try to find the bug that caused the infinite loop on the game I submitted, but struggled quite a bit reading the lex code. So seeing #77 I took a stab at refactoring most of the logic to make it more readable for myself. I appreciate that readability can be quite subjective, so let me know if this is too extensive, otherwise I'd be happy to make adaptations and suggestions you provide. All existing tests are passing, with just some changes to invocations. I did a crash course on lemke-howson, so not unlikely I got some of the elements wrong.

As for the actual bug, it should be possible to backport this without the refactoring. See comment in code.

### CODE REVIEW

File: src/nashpy/linalg/tableau.py [hunk 1]
```diff
@@ -0,0 +1,492 @@
...
+            The rows to perform tiebreak on
+
+        Returns
+        -------
+        int
+            The row to pivot on
+        """
+        errs = self._tableau[:, sorted(self.slack_variables)]
+        pivot_column = self._tableau[:, (column_index,)]
+        err_ratios = errs / pivot_column
```
Comment (tokheim) [inline src/nashpy/linalg/tableau.py:454]: If you compare with https://github.com/drvinceknight/Nashpy/pull/209/files#diff-afcbc2dd5fe0e1b292a6d2b26086885e45cbd461a859d4a32100904e20811b0eL50, this seems to be the fix for the bug. The original lex sort did not take into account the pivot column when lex sorting the perturbed tableau.

Beyond this had to add some handling for NaNs. Also struggled a bit getting test cases to complete. Basically its not entirely clear to me why we can do `pivot_col / tableau[:,-1]` in the max ratio check, while here its swapped.

File: tests/unit/test_lemke_howson_lex.py [hunk 2]
```diff
@@ -76,16 +78,16 @@ def test_lemke_howson_lex_degenerate_tie_breaking_looping(
...
+                if not (np.isnan(eq[0]).any() or np.isnan(eq[1]).any()):
+                    nonnans += 1
+                else:
+                    continue
                 reward = eq[0].dot(A).dot(eq[1].transpose())
                 self.assertAlmostEqual(reward, expected_reward, delta=1e-7)
-                found_eq = True
-            self.assertTrue(found_eq, "Did not find eq on label " + str(label))
+        self.assertGreaterEqual(
+            nonnans, 14, msg="at least 14 eqs without nan values produced"
```
Comment (tokheim) [inline tests/unit/test_lemke_howson_lex.py:92]: Even with lex sorting working, some of the labels generate NaNs. It would seem like this has to do with the pivoting algorithm sometimes producing high row values after many iterations and numerical instability. I haven't looked much further into this

**Comment by drvinceknight** [comment]
Thanks a lot for this @tokheim. I'll have time over the next week or so to review.

File: src/nashpy/algorithms/lemke_howson.py [hunk 3]
```diff
@@ -2,73 +2,16 @@
...
+from typing import Tuple
+from nashpy.linalg import TableauBuilder
 
 
 def lemke_howson(
-    A: npt.NDArray, B: npt.NDArray, initial_dropped_label: int = 0
+    A: npt.NDArray,
+    B: npt.NDArray,
+    initial_dropped_label: int = 0,
+    algorithm: str = "basic",
```
Comment (drvinceknight) [inline src/nashpy/algorithms/lemke_howson.py:14]: What do you think about changing this to:

```suggestion
    lexicographic: bool = False,
```

I'm wondering however if perhaps it might be worth just having lexicographic as the default -- in which case perhaps there's no need for the non lexicographic Tableau at all. What do you think?

File: src/nashpy/linalg/tableau.py [hunk 4]
```diff
@@ -0,0 +1,492 @@
+"""A class for the tableaus used in the Lemke Howson algorithms"""
+import warnings
+import numpy as np
+import numpy.typing as npt
+from typing import Set, List, Iterable, Optional
+
+
+class TableauBuilder(object):
```
Comment (drvinceknight) [inline src/nashpy/linalg/tableau.py:8]: Given the use case, could this be a function instead of a class?

File: tests/unit/test_tableau.py
Comment (drvinceknight) [inline tests/unit/test_tableau.py:1]: We should change the module doctoring here and I think I'd like to see more unit tests for each of the methods in the class (but we can do that in a separate PR).

**Review by drvinceknight — CHANGES_REQUESTED** [review]
This looks really good. I really like the refactor to the Tableu class. I did drop a question/thought that I'd appreciate your thoughts on.

File: src/nashpy/algorithms/lemke_howson.py (hunk 3 above)
Comment (tokheim) [reply src/nashpy/algorithms/lemke_howson.py:14]: I think there are a few questions packed into one here, will try to answer them all seperately

1. should non-lexiographic tableau exist at all:
I understand that part of the purpose of the library is being a teaching tool. From that perspective I think it really makes sense to offer both, as the lexiographic class sort of highlights that these are the few places you'd adapt lemke howson in order to handle degenerate matrices. At least for me just understanding the basic algorithm was already complex enough, and the theory around lexiographic sorting was a very cryptic layer on top of that again. So offering lex sort as an extension to the normal tableau makes the differences very explicit.
2. what should the default/options be when calling `Game::lemke_howson_enumeration``:
I took the conservative approach of keeping the behavior exactly the same as today. I probably think you should decide what the most sensible defaults and options are. I randomly stumbled into nash equilibriums and most of my cases were degenerate games. So I would maybe think that given the performance is quite similar but lex-sorting more versatile, I would lean towards at least default being lex sorting.
3. String algorithm, vs bool lexiographic option
I drew inspiration from https://docs.scipy.org/doc/scipy/tutorial/optimize.html when going for string. Still I agree bool makes sense if there are just two different implementations. If you think that is the case, I'd be happy to make that change, though potentially I see there might be more algorithms like:
* until I managed to solve the lex sorting bug I considered experimenting with the cycle-breaking approach of picking a different row to pivot each time we had the same basic variables and same label to drop
* could there be other pivoting algorithms (potentially with or without lex sorting). What we do today maybe emphasizes numerical and integer stability, but could run into infs and nans on large matrices. Potentially there could be an algorithm that handles those types of matrices better.

File: src/nashpy/linalg/tableau.py (hunk 4 above)
Comment (tokheim) [reply src/nashpy/linalg/tableau.py:8]: I agree the builder feels a bit strange so happy to change here. I went into that direction to avoid a single method with too many parameters. Still if you wanted to change it would you want:

1. a single createRowTableau/createColumnTableau method with parameters for payoffs and algorithm
2. that lemke howson classs more calls individual methods to first create tableau matrix, and then possibly roll it and finally generate tableau class?

File: src/nashpy/algorithms/lemke_howson.py (hunk 3 above)
Comment (drvinceknight) [reply src/nashpy/algorithms/lemke_howson.py:14]: Thank you very much for your thoughtful response.

> should non-lexiographic tableau exist at all:
> I understand that part of the purpose of the library is being a teaching tool. From that perspective I think it really makes sense to offer both, as the lexiographic class sort of highlights that these are the few places you'd adapt lemke howson in order to handle degenerate matrices. At least for me just understanding the basic algorithm was already complex enough, and the theory around lexiographic sorting was a very cryptic layer on top of that again. So offering lex sort as an extension to the normal tableau makes the differences very explicit.

I agree. This will also help when I refactor the documentation on LH.

> what should the default/options be when calling `Game::lemke_howson_enumeration``:
I took the conservative approach of keeping the behavior exactly the same as today. I probably think you should decide what the most sensible defaults and options are. 

I appreciate the way you've put this PR together greatly :)

> I randomly stumbled into nash equilibriums and most of my cases were degenerate games. So I would maybe think that given the performance is quite similar but lex-sorting more versatile, I would lean towards at least default being lex sorting.

Let's do it! If you could make the change to make lex the default that'd be great.

> String algorithm, vs bool lexiographic option
I drew inspiration from https://docs.scipy.org/doc/scipy/tutorial/optimize.html when going for string. Still I agree bool makes sense if there are just two different implementations. If you think that is the case, I'd be happy to make that change, though potentially I see there might be more algorithms like:
until I managed to solve the lex sorting bug I considered experimenting with the cycle-breaking approach of picking a different row to pivot each time we had the same basic variables and same label to drop
could there be other pivoting algorithms (potentially with or without lex sorting). What we do today maybe emphasizes numerical and integer stability, but could run into infs and nans on large matrices. Potentially there could be an algorithm that handles those types of matrices better.

Let's go for bool, I think we're on the same page here. I don't want to over optimise for future changes though so let's just go with what's nice now.

File: tests/unit/test_tableau.py
Comment (drvinceknight) [reply tests/unit/test_tableau.py:1]: > To make the change less scary I tried to keep the test classes as similar as possible, but do let me know if we should merge lex and regular test cases into one file etc.

Please do merge :)

> I'd appreciate leaving more test cases as a separate PR, simply because I'm not sure how many high quality test cases I could provide myself.

100% I'll open an issue when we merge this PR.

**Comment by drvinceknight** [comment]
Thanks again for the awesome work here. Looking forward to merging this contribution.

File: src/nashpy/algorithms/lemke_howson.py (hunk 3 above)
Comment (tokheim) [reply src/nashpy/algorithms/lemke_howson.py]: Switched to lexiographic boolean and default is on

File: src/nashpy/linalg/tableau.py (hunk 4 above)
Comment (tokheim) [reply src/nashpy/linalg/tableau.py]: Updated, think this is a readability improvement as well

File: src/nashpy/algorithms/lemke_howson_lex.py [hunk 5]
```diff
@@ -1,14 +1,15 @@
...
+from deprecated import deprecated  # type: ignore
 import numpy.typing as npt
 from typing import Tuple
-from nashpy.integer_pivoting import make_tableau, pivot_tableau_lex
 
-from .lemke_howson import shift_tableau, tableau_to_strategy
+from .lemke_howson import lemke_howson
 
 
+@deprecated(
```
Comment (tokheim) [inline src/nashpy/algorithms/lemke_howson_lex.py:9]: We never discussed what to do with the old lex method. I left it in, with a deprecation warning in case its still used by some

**Comment by drvinceknight** [comment]
> requested a review from [drvinceknight](https://github.com/drvinceknight)

It's been marking season over here so I'm coming up for air finally thanks for your patience.
